src/lakefs_spec/transaction.py

```diff
@@ -0,0 +1,144 @@
+        self.value = value
+
+    def unwrap(self) -> T:
+        if self.value is None:
+            raise RuntimeError("placeholder unfilled")
+        return self.value
+
+
+def unwrap_placeholders(kwargs: dict[str, Any]) -> dict[str, Any]:
+    return {k: v.unwrap() if isinstance(v, Placeholder) else v for k, v in kwargs.items()}
+
+
+class LakeFSTransaction(Transaction):
+    """A lakeFS transaction model capable of versioning operations in between file uploads."""
+
+    def __init__(self, fs: "LakeFSFileSystem"):
+        """
+        Initialize a lakeFS transaction. The base class' `file` stack can also contain
+        versioning operations.
+        """
+        super().__init__(fs=fs)
+        self.fs: "LakeFSFileSystem"
+        self.files: deque[AbstractBufferedFile | VersioningOpTuple] = deque(self.files)
+
+    def __enter__(self):
+        self.fs._intrans = True
+        return self
+
+    def commit(
+        self, repository: str, branch: str, message: str, metadata: dict[str, str] | None = None
+    ) -> Placeholder[Commit]:
+        """
+        Create a commit on a branch in a repository with a commit message and attached metadata.
+        """
+
+        # bind all arguments to the client helper function, and then add it to the file-/callstack.
+        op = partial(
+            commit, repository=repository, branch=branch, message=message, metadata=metadata
+        )
+        p: Placeholder[Commit] = Placeholder()
+        self.files.append((op, p))
+        # return a placeholder for the commit.
+        return p
+
+    def complete(self, commit: bool = True) -> None:
+        """
+        Finish transaction: Unwind file+versioning op stack via
+         1. Committing or discarding in case of a file, and
+         2. Conducting versioning operations using the file system's client.
+
+         No operations happen and all files are discarded if `commit` is False,
+         which is the case e.g. if an exception happens in the context manager.
+        """
+        while self.files:
+            # fsspec base class calls `append` on the file, which means we
+            # have to pop from the left to preserve order.
+            f = self.files.popleft()
+            if isinstance(f, AbstractBufferedFile):
+                if commit:
+                    f.commit()
+                else:
+                    f.discard()
+            else:
+                # client helper + return value case.
+                op, retval = f
+                if commit:
+                    result = op(self.fs.client)
+                    # if the transaction member returns a placeholder,
+                    # fill it with the result of the client helper.
+                    if isinstance(retval, Placeholder):
+                        retval.set_value(result)
+
+        self.fs._intrans = False
+
+    def create_branch(self, repository: str, branch: str, source_branch: str) -> str:
+        op = partial(
```

Comment:
```suggestion
    def create_branch(self, repository: str, branch: str, source_branch: str) -> str:
    """Create a new branch from source branch in a repository."""
        op = partial(
```

Reply:
Added the docstrings for `create_branch` and `rev_parse` in [f9c7ba6](https://github.com/appliedAI-Initiative/lakefs-spec/pull/140/commits/f9c7ba6e48b3182259339c081203f52a0e1b5325)